### PR TITLE
Added support for `loadDartLibrarySources` VM service RPC.

### PR DESCRIPTION
**Pull request by bkonyi**
`loadDartLibrarySources` allows for dart:* sources to be loaded from an unstripped platform_strong.dill, making debugging with dart:* sources possible. This change allows for the VM to delegate loading the unstripped kernel file to flutter_tools on the host before processing the loaded kernel on the device.

Depends on [this CL](https://dart-review.googlesource.com/c/sdk/+/92407).

### CODE REVIEW

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 1]
```diff
@@ -356,6 +369,19 @@ class VMService {
     }
   }
 
+  Future<Map<String, dynamic>> _handleLoadDartLibrarySources(Map<String,
+      dynamic> data) async {
+    final path = data['kernelPath'];
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/vmservice.dart]: I'd add types here (`final String path`), and below.

Comment (devoncarew) [reply packages/flutter_tools/lib/src/vmservice.dart]: also, name it `kernelPath`?

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 2]
```diff
@@ -356,6 +369,19 @@ class VMService {
...
+    final bytes = await fs.file(path).readAsBytes();
+    final result = Base64Encoder().convert(bytes);
+    return <String, dynamic>{
+      'type': 'Success',
+      'result': <String, String> {
+        'kernelBytes': result
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/vmservice.dart]: add a trailing comma?

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 3]
```diff
@@ -356,6 +369,19 @@ class VMService {
     }
   }
 
+  Future<Map<String, dynamic>> _handleLoadDartLibrarySources(Map<String,
+      dynamic> data) async {
+    final path = data['kernelPath'];
+    final bytes = await fs.file(path).readAsBytes();
+    final result = Base64Encoder().convert(bytes);
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/vmservice.dart]: Can you use the top-level dart:convert `base64` variable here (instead of creating a new instance)?

**Review by devoncarew — APPROVED** [review]
lgtm w/ some notes in-line

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 4]
```diff
@@ -114,6 +115,18 @@ class VMService {
...
+    _peer.registerMethod('loadDartLibrarySources',
+        (rpc.Parameters params) async {
+      return await _handleLoadDartLibrarySources(params.asMap);
+    });
+
+    // If the Flutter Engine doesn't support service registration this will
+    // have no effect
+    _peer.sendNotification('_registerService', <String, String> {
+      'service': 'loadDartLibrarySources',
+      'alias': 'Flutter Tools',
```
[the remark this answers is not in the document]
Comment (bkonyi) [reply packages/flutter_tools/lib/src/vmservice.dart]: Done.

**Review by a-siva — APPROVED** [review]

File: packages/flutter_tools/lib/src/vmservice.dart [hunk 5]
```diff
@@ -378,17 +377,28 @@ class VMService {
...
-      'type': 'Success',
-      'result': <String, String> {
-        'kernelBytes': result,
-      },
-    };
+  Future<Map<String, dynamic>> _handleLoadDartLibrarySources() async {
+    // We assume that the current program is being run using the dart executable
+    // from the dart-sdk packaged with Flutter.
+    const relativeKernelPath =
+      '../../artifacts/engine/common/flutter_patched_sdk/platform_strong.dill';
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/vmservice.dart]: I suspect that there's a way to query the flutter_tools api for the artifacts location (possibly via the `Artifacts` class)? cc @jonahwilliams @aam

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/vmservice.dart]: Something like `artifacts.getArtifactPath(Artifact. platformKernelDill)`

Comment (bkonyi) [reply packages/flutter_tools/lib/src/vmservice.dart]: Good call, totally forgot that existed. Done.

**Comment by bkonyi** [comment]
Abandoning this change as we've decided to move forward with linking in sources for Flutter debug builds (see [this CL](https://dart-review.googlesource.com/c/sdk/+/93375) for relevant Dart SDK changes; Flutter changes coming).

**Pull request closed without merging**